### PR TITLE
MOD-2518: Use REDISMODULE_EVENT_SHUTDOWN to clear resources

### PR DESCRIPTION
**Pull request by ashtul**
* Using a server notification might add stability to clearing resources as a server shuts down.
Currently, we use `__attribute__((destructor))` which is called when the shared library is unloaded and we see some instability.

* Use a local `keepalive` variable to stop threadpool instead of a global one. It prevents `thpool_wait` from freezing if called after `thpool_destroy` was called on another pool.

### CODE REVIEW

**Review by rafie — DISMISSED** [review]

File: src/notifications.c [hunk 1]
```diff
@@ -296,6 +296,16 @@ void ShardingEvent(RedisModuleCtx *ctx, RedisModuleEvent eid, uint64_t subevent,
   }
 }
 
+
+void ShutdownEvent(RedisModuleCtx *ctx, RedisModuleEvent eid, uint64_t subevent, void *data) {
+  if (eid.id != REDISMODULE_EVENT_SHUTDOWN) {
+    RedisModule_Log(RSDummyContext, "warning", "Bad event given, ignored.");
```
Comment (oshadmi) [inline src/notifications.c:302]: Seems we can just silently skip non-interesting events and avoid logging them as "Bad"

File: src/notifications.c [hunk 2]
```diff
@@ -313,6 +323,12 @@ void Initialize_KeyspaceNotifications(RedisModuleCtx *ctx) {
       RedisModule_SubscribeToServerEvent(ctx, RedisModuleEvent_Sharding, ShardingEvent);
     }
   }
+
+  if (RedisModule_SubscribeToServerEvent && getenv("RS_GLOBAL_DTORS")) {
+    // clear resources when the server exits
```
Comment (oshadmi) [inline src/notifications.c:328]: What about the comment `// used only with sanitizer or valgrind` ?

**Review by oshadmi — COMMENTED** [review]
👍🏼 Small comments

**Review by rafie — APPROVED** [review]

**Review by MeirShpilraien — APPROVED** [review]

**Review by oshadmi — APPROVED** [review]
👍🏼

**Pull request merged**